### ISSUE TITLE
[C++] Compute: KeyEncoder support List Type

### ISSUE THREAD

**Issue by mapleFU** [apache/arrow#43911]
### Describe the enhancement requested

This is part of https://github.com/apache/arrow/issues/43716 .

We need first support KeyEncoder in ListType

### Component(s)

C++

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 365 days. Please remove the stale label or comment below, or this issue will be closed in 14 days. If this improvement is still desired but has no current owner, please add the 'Status: needs champion' label.

**Comment by dangotbanned** [comment]
I think this has a closing PR from last month?

- [ ] #47875

**Comment by thisisnic** [comment]
Good point, the CI job to link them doesn't differentiate, I'll remove the label.  Thanks for flagging!
